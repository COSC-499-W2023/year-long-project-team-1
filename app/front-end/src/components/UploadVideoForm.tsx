"use client";
import { useState } from "react";
import { JSONResponse } from "@lib/json";
import { Button, Stack, StackItem, Grid, GridItem } from "@patternfly/react-core";

export const UploadVideoForm = () => {
    const [file, setFile] = useState<File>();
    const [filename, setFilename] = useState<string>("");
    const [isPicked, setIsPicked] = useState<boolean>(false);
    const [responseData, setResponseData] = useState<JSONResponse>();
    const acceptedMimeTypes = ["video/mp4", "video/x-msvideo", "video/quicktime"]; // mp4, avi, mov

    const onSubmitClick = async (e: any) => {
        if (!file || !isPicked) {
            alert("No file selected!");
            return;
        }

        try {
            const formData = new FormData();
            formData.set("file", file);

            const response = await fetch("/api/video/upload", {
                method: "POST",
                body: formData,
            });

            const json = await response.clone().json();

            if (!response.ok) {
                console.error("Error in upload response.");
                throw Error(await response.text());
            }

            setResponseData(json);
        } catch (err: any) {
            console.error(err.message);
        }
    };

    const onFileChanged = (e: any) => {
        const f = e.target.files?.[0];
        if (!acceptedMimeTypes.includes(f.type)) {
            alert("You must select an *.mp4, *.avi, or *.mov file");
            return;
        }
        setFile(f);
        setIsPicked(true);
        setFilename(f.name);
    };

    return (
        <Stack>
            <StackItem>
                {/* <FileUpload
					id="videoupload"
					filename={filename}
					onFileInputChange={(e: DropEvent, f: File) => {
						setFile(f);
						setIsPicked(true);
						setFilename(f.name);
					}}
					onClearClick={() => {
						setFile(undefined);
						setFilename("");
						setIsPicked(false);
					}}
					dropzoneProps={{
						accept: {
							"video/mp4": [".mp4"],
							"video/x-msvideo": [".avi"],
							"video/quicktime": [".mov"],
						},
						onDropRejected: (
							fileRejectionList,
							event: DropEvent
						) => {
							alert(
								"Invalid file type! Must be a *.mp4, *.avi, or *.mov file."
							);
							// setFile(undefined);
							setFilename("");
							setIsPicked(false);
						},
					}}
				/> */}
                <input id="videoupload" type="file" accept={acceptedMimeTypes.toString()} onChange={onFileChanged} />
                <text>{filename}</text>
            </StackItem>
            <StackItem>
                <Grid>
                    <GridItem span={4}></GridItem>
                    <GridItem span={4}>
                        <Button
                            variant="danger"
                            onClick={(e) => {
                                alert("Not implemented yet!");
                            }}>
                            Record video
                        </Button>
                    </GridItem>
                    <GridItem span={4}>
                        <Button variant="primary" onClick={onSubmitClick}>
                            Submit video
                        </Button>
                    </GridItem>
                </Grid>
            </StackItem>
        </Stack>
    );
};
export default UploadVideoForm;
