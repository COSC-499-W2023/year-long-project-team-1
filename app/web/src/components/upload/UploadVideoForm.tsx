/*
 * Copyright [2023] [Privacypal Authors]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
"use client";
import { useEffect, useRef, useState } from "react";
import { JSONResponse } from "@lib/response";
import {
  Button,
  Card,
  CardTitle,
  CardBody,
  ActionList,
  ActionListItem,
  Form,
  Radio,
} from "@patternfly/react-core";
import { useRouter } from "next/navigation";
import style from "@assets/style";
// https://github.com/DeltaCircuit/react-media-recorder/issues/105
// was having a strange bug with this, but someone made a version
// specifically to fix the bug since the maintainers weren't fixing them
import { useReactMediaRecorder } from "react-media-recorder-2";

export const UploadVideoForm = () => {
  const router = useRouter();

  const [uploadChecked, setUploadChecked] = useState<boolean>(true);
  const handleUploadChanged = () => {
    setUploadChecked(!uploadChecked);
  };
  const onRecordStop = (blobUrl: string, blob: Blob) => {
    const f = new File([blob], "recorded.mp4", { type: "video/mp4" });
    setRecordFile(f);
  };

  const [localFile, setLocalFile] = useState<File>();
  const [recordFile, setRecordFile] = useState<File>();
  const [file, setFile] = useState<File>();
  const [isPicked, setIsPicked] = useState<boolean>(false);
  const [responseData, setResponseData] = useState<JSONResponse>();
  const acceptedMimeTypes = ["video/mp4", "video/x-msvideo", "video/quicktime"]; // mp4, avi, mov
  const { status, startRecording, stopRecording, mediaBlobUrl, previewStream } =
    useReactMediaRecorder({
      video: { frameRate: 24 },
      blobPropertyBag: { type: "video/mp4" },
      onStop: onRecordStop,
    }); // force a lower but still standard fps to improve performance

  const onSubmitClick = async (e: any) => {
    if (uploadChecked) {
      setFile(localFile);
    } else {
      setFile(recordFile);
    }
    // these ^ aren't updating before this if check here? but only the first time, the next submit click it works just fine (????)
    if (!file || (!isPicked && uploadChecked)) {
      console.log(uploadChecked, file, localFile, recordFile);
      alert(
        "No file selected. Make sure you either upload or record a video and select the correct upload type.",
      );
      return;
    }

    try {
      const formData = new FormData();
      formData.set("file", file);

      const response = await fetch("/api/video/upload", {
        method: "POST",
        body: formData,
      });

      if (!response.ok) {
        console.error("Error in upload response.");
        throw Error(await response.text());
      } else if (response.status === 200) {
        const json = await response.clone().json();

        setResponseData(json);

        setTimeout(() => {
          router.push(
            `/upload/status/${encodeURIComponent(json.data?.filePath)}`,
          );
        }, 150);
      }
    } catch (err: any) {
      console.error(err.message);
    }
  };

  const onRecordClick = (e: any) => {
    if (status !== "recording") {
      startRecording();
    } else {
      stopRecording();
    }
  };

  const onFileChanged = (e: any) => {
    const f = e.target.files?.[0] as File;
    if (!acceptedMimeTypes.includes(f.type)) {
      alert("You must select an *.mp4, *.avi, or *.mov file");
      return;
    }
    setLocalFile(f);
    setIsPicked(true);
  };

  const LiveFeed = ({ stream }: { stream: MediaStream | null }) => {
    const ref = useRef<HTMLVideoElement>(null);
    useEffect(() => {
      if (ref.current && stream) {
        ref.current.srcObject = stream;
      }
    }, [stream]);
    return stream ? <video ref={ref} autoPlay /> : null;
  };

  return (
    <Card style={style.card} aria-label="Video uploader">
      <CardTitle component="h1">Upload a Video</CardTitle>
      <CardBody>
        {responseData?.data?.success ? (
          <p className="success">Upload successful. Redirecting...</p>
        ) : (
          <Form
            aria-label="Video upload form"
            onSubmit={(e) => e.preventDefault()}
          >
            <input
              className="file-input"
              type="file"
              alt="file upload"
              accept={acceptedMimeTypes.toString()}
              onChange={onFileChanged}
            />
            {status === "stopped" ? (
              <video src={mediaBlobUrl} controls />
            ) : null}
            {status === "recording" ? (
              <LiveFeed stream={previewStream} />
            ) : null}
            <ActionList style={style.actionList}>
              <ActionListItem>
                <Radio
                  isChecked={uploadChecked}
                  onChange={handleUploadChanged}
                  name="uploadType"
                  id="upload"
                  label="Uploaded"
                />
              </ActionListItem>
              <ActionListItem>
                <Radio
                  isChecked={!uploadChecked}
                  onChange={handleUploadChanged}
                  name="uploadType"
                  id="record"
                  label="Recorded"
                />
              </ActionListItem>
              <ActionListItem>
                <Button
                  variant="primary"
                  type="submit"
                  onClick={onSubmitClick}
                  aria-label="Submit video"
                >
                  Submit video
                </Button>
              </ActionListItem>
              <ActionListItem>
                <Button
                  variant="danger"
                  onClick={onRecordClick}
                  aria-label="Record video"
                >
                  {status !== "recording" ? "Record video" : "Stop recording"}
                </Button>
              </ActionListItem>
            </ActionList>
          </Form>
        )}
      </CardBody>
    </Card>
  );
};
export default UploadVideoForm;
