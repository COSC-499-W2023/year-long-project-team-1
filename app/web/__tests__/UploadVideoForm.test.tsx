import { render, screen, fireEvent } from "@testing-library/react";
import UploadVideoForm from "@components/upload/UploadVideoForm";
import "@testing-library/jest-dom";

describe("Page", () => {
    it("UploadVideoForm matches snapshot", () => {
        const snap = render(<UploadVideoForm />);
        expect(snap).toMatchSnapshot();
    });

    it("UploadVideoForm contains both buttons", () => {
        render(<UploadVideoForm />);
        expect(screen.getByRole("button", { name: "Record video" })).toBeInTheDocument();
        expect(screen.getByRole("button", { name: "Submit video" })).toBeInTheDocument();
    });

    it("UploadVideoForm has a file input component", () => {
        render(<UploadVideoForm />);
        expect(screen.getByAltText("file upload")).toBeInTheDocument();
    });
});
