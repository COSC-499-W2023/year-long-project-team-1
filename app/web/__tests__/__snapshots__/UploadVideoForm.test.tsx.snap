// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page UploadVideoForm matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="pf-v5-c-card"
        data-ouia-component-id="OUIA-Generated-Card-1"
        data-ouia-component-type="PF5/Card"
        data-ouia-safe="true"
        id=""
      >
        <div
          class="pf-v5-c-card__title"
        >
          <h1
            class="pf-v5-c-card__title-text"
          >
            Upload a Video
          </h1>
        </div>
        <div
          class="pf-v5-c-card__body"
        >
           
          <input
            accept="video/mp4,video/x-msvideo,video/quicktime"
            alt="file upload"
            class="file-input"
            type="file"
          />
          <div
            class="pf-v5-c-action-list"
            style="display: flex; flex-direction: row; justify-content: center;"
          >
            <div
              class="pf-v5-c-action-list__item"
            >
              <button
                aria-disabled="false"
                class="pf-v5-c-button pf-m-primary"
                data-ouia-component-id="OUIA-Generated-Button-primary-1"
                data-ouia-component-type="PF5/Button"
                data-ouia-safe="true"
                type="button"
              >
                Submit video
              </button>
            </div>
            <div
              class="pf-v5-c-action-list__item"
            >
              <button
                aria-disabled="true"
                class="pf-v5-c-button pf-m-danger pf-m-disabled"
                data-ouia-component-id="OUIA-Generated-Button-danger-1"
                data-ouia-component-type="PF5/Button"
                data-ouia-safe="true"
                disabled=""
                type="button"
              >
                Record video
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="pf-v5-c-card"
      data-ouia-component-id="OUIA-Generated-Card-1"
      data-ouia-component-type="PF5/Card"
      data-ouia-safe="true"
      id=""
    >
      <div
        class="pf-v5-c-card__title"
      >
        <h1
          class="pf-v5-c-card__title-text"
        >
          Upload a Video
        </h1>
      </div>
      <div
        class="pf-v5-c-card__body"
      >
         
        <input
          accept="video/mp4,video/x-msvideo,video/quicktime"
          alt="file upload"
          class="file-input"
          type="file"
        />
        <div
          class="pf-v5-c-action-list"
          style="display: flex; flex-direction: row; justify-content: center;"
        >
          <div
            class="pf-v5-c-action-list__item"
          >
            <button
              aria-disabled="false"
              class="pf-v5-c-button pf-m-primary"
              data-ouia-component-id="OUIA-Generated-Button-primary-1"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe="true"
              type="button"
            >
              Submit video
            </button>
          </div>
          <div
            class="pf-v5-c-action-list__item"
          >
            <button
              aria-disabled="true"
              class="pf-v5-c-button pf-m-danger pf-m-disabled"
              data-ouia-component-id="OUIA-Generated-Button-danger-1"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe="true"
              disabled=""
              type="button"
            >
              Record video
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
